src/MySqlConnector/MySqlClient/MySqlDataReader.cs

```diff
@@ -781,12 +830,32 @@ private enum State
 			NoMoreData,
 		}
 
+	    public async Task WaitDoneReadingAsync(CancellationToken cancellationToken=default(CancellationToken))
+	    {
+	        cancellationToken.ThrowIfCancellationRequested();
+	        await m_done_reading.WaitAsync(cancellationToken).ConfigureAwait(false);
+	        m_done_reading.Release();
+	    }
+
 		static readonly Task<bool> s_falseTask = Task.FromResult(false);
 		static readonly Task<bool> s_trueTask = Task.FromResult(true);
 
 		MySqlCommand m_command;
 		MySqlSession m_session;
-		State m_state;
+	    SemaphoreSlim m_done_reading = new SemaphoreSlim(1);
+	    State _m_state;
+	    State m_state
```

Comment:
Shouldn't have two fields with almost identical names. This could become a private property `State`, or a `private SetState(State state)` method could be added to encapsulate the setting logic.